File: permutate.py



def permutate(word):
    permutations = []

    word_list = [ch for ch in word]
    word_length = len(word)
    
    def permutate_recurse(x, leftover):
        if len(leftover) == 0:
            return x

        for i in range(0, len(leftover)):
            x_temp = x
            leftover_temp = list(leftover)
            char = leftover[i]
            del leftover_temp[i]
            
            if len(x_temp + char) == word_length:
            	permutations.append(x_temp + char)

            permutate_recurse(x_temp + char, leftover_temp)

    permutate_recurse('', word_list)
    return permutations

print permutate('abc')

File: scala/permutate.scala
 

def permutate(word: String): List[String] = {
	var permutations = List[String]()
	val wordList = word.toString.split("").toList
	val wordLength = wordList.length

    def permutateRecurse(x: String, leftover: List[String]): List[String] = {
        if(leftover.length == 0) return Nil

        for(i <- 0 to leftover.length - 1) {
            var leftoverTmp = leftover
            val (left, right) = leftoverTmp.splitAt(i)
            leftoverTmp = left ::: right.drop(1)
            val permValue = x + leftover(i)
            
            if(permValue.length == wordLength) permutations = permutations ::: List(permValue)
            permutateRecurse(permValue, leftoverTmp)
        }
        return Nil
    }
    permutateRecurse("",  word.toString.split("").toList)
    return permutations
}

val test = permutate("dustin")
println(test.length)
